File: aletheia/rag/storage/vector_index.py
"""
Milvus Vector Index - Semantic Search Layer
Stores sentence embeddings for approximate nearest neighbor search.
Uses MilvusClient for Zilliz Cloud compatibility.
"""

import requests
from pymilvus import MilvusClient, DataType
from typing import List, Dict, Optional
from openai import OpenAI

from aletheia.config import milvus_config, embedding_config


class VectorIndex:
    """Milvus-based vector search with metadata filtering using MilvusClient."""

    def __init__(
        self,
        collection_name: str = "sentence_embeddings",
        uri: str = None,
        token: str = None,
    ):
        """
        Initialize Milvus connection and collection.

        Args:
            collection_name: Name of the Milvus collection
            uri: Milvus URI (uses config if None). For Zilliz Cloud: https://xxx.zillizcloud.com:19530
            token: Authentication token (uses config if None). Format: "user:password"
        """
        self.collection_name = collection_name
        self.dimension = embedding_config.dimension
        self.embedding_model = embedding_config.model

        # Initialize embedding client based on provider
        self.provider = embedding_config.provider
        if self.provider == "openai":
            self.openai_client = OpenAI(api_key=embedding_config.openai_api_key)
        elif self.provider == "ollama":
            self.openai_client = None  # Not used for Ollama
            self.ollama_base_url = embedding_config.ollama_base_url
        else:
            raise ValueError(f"Unknown embedding provider: {self.provider}")

        # Get connection parameters
        if uri or token:
            # Use provided parameters
            self.uri = uri or milvus_config.uri
            self.token = token or milvus_config.token
        else:
            # Use config
            conn_params = milvus_config.connection_params
            self.uri = conn_params["uri"]
            self.token = conn_params.get("token", "")

        # Connect to Milvus using MilvusClient
        self._connect()

    def _connect(self):
        """Establish connection to Milvus using MilvusClient."""
        try:
            if self.token:
                self.client = MilvusClient(uri=self.uri, token=self.token)
                print(f"✓ Connected to Zilliz Cloud at {self.uri}")
            else:
                self.client = MilvusClient(uri=self.uri)
                print(f"✓ Connected to Milvus at {self.uri}")
        except Exception as e:
            raise ConnectionError(f"Failed to connect to Milvus: {e}")

    def create_collection(self, drop_existing: bool = False):
        """
        Create Milvus collection with schema using MilvusClient.

        Args:
            drop_existing: If True, drop existing collection before creating
        """
        # Drop existing collection if requested
        if drop_existing and self.client.has_collection(self.collection_name):
            self.client.drop_collection(self.collection_name)
            print(f"✓ Dropped existing collection: {self.collection_name}")

        # Check if collection already exists
        if self.client.has_collection(self.collection_name):
            print(f"✓ Collection already exists: {self.collection_name}")
            return

        # Define schema using MilvusClient API
        schema = self.client.create_schema(auto_id=False, enable_dynamic_field=False)

        # Add fields
        schema.add_field(
            field_name="id", datatype=DataType.VARCHAR, max_length=200, is_primary=True
        )
        schema.add_field(
            field_name="embedding", datatype=DataType.FLOAT_VECTOR, dim=self.dimension
        )
        schema.add_field(field_name="doc_id", datatype=DataType.VARCHAR, max_length=100)
        schema.add_field(field_name="page_num", datatype=DataType.INT64)
        schema.add_field(
            field_name="paragraph_id", datatype=DataType.VARCHAR, max_length=100
        )
        schema.add_field(
            field_name="sentence_id", datatype=DataType.VARCHAR, max_length=100
        )

        # Create index params
        index_params = self.client.prepare_index_params()
        index_params.add_index(
            field_name="embedding",
            index_type="IVF_FLAT",
            metric_type="COSINE",
            params={"nlist": 1024},
        )

        # Create collection
        self.client.create_collection(
            collection_name=self.collection_name,
            schema=schema,
            index_params=index_params,
        )

        print(f"✓ Created collection: {self.collection_name} with COSINE similarity")

    def _generate_embeddings(self, texts: List[str]) -> List[List[float]]:
        """
        Generate embeddings using configured provider (OpenAI or Ollama).

        Args:
            texts: List of text strings

        Returns:
            List of embedding vectors
        """
        try:
            if self.provider == "ollama":
                return self._generate_ollama_embeddings(texts)
            else:
                return self._generate_openai_embeddings(texts)
        except Exception as e:
            raise Exception(f"Failed to generate embeddings: {e}")

    def _generate_openai_embeddings(self, texts: List[str]) -> List[List[float]]:
        """Generate embeddings using OpenAI API."""
        from tenacity import (
            retry,
            stop_after_attempt,
            wait_exponential,
            retry_if_exception_type,
        )

        @retry(
            retry=retry_if_exception_type(Exception),
            stop=stop_after_attempt(5),
            wait=wait_exponential(multiplier=1, min=2, max=20),
            reraise=True,
        )
        def _create_embeddings():
            return self.openai_client.embeddings.create(
                model=self.embedding_model, input=texts
            )

        response = _create_embeddings()
        return [item.embedding for item in response.data]

    def _generate_ollama_embeddings(self, texts: List[str]) -> List[List[float]]:
        """Generate embeddings using Ollama HTTP API with batch processing."""
        from aletheia.rag.processing.batch_embeddings import BatchEmbeddingClient

        # Use batch embedding client for better performance
        client = BatchEmbeddingClient(
            provider="ollama",
            model=self.embedding_model,
            base_url=self.ollama_base_url,
            batch_size=32,
            max_workers=4,
        )

        try:
            embeddings = client.embed_batch(texts, batch_size=32)
            return embeddings
        except Exception as e:
            print(f"  ⚠️ Batch embedding failed: {e}, falling back to sequential")
            # Fallback to original sequential method
            return self._generate_ollama_embeddings_sequential(texts)
        finally:
            client.close()

    def _generate_ollama_embeddings_sequential(
        self, texts: List[str]
    ) -> List[List[float]]:
        """Fallback sequential embedding generation."""
        from tenacity import (
            retry,
            stop_after_attempt,
            wait_exponential,
            retry_if_exception_type,
        )

        embeddings = []

        @retry(
            retry=retry_if_exception_type((requests.RequestException, Exception)),
            stop=stop_after_attempt(5),
            wait=wait_exponential(multiplier=1, min=2, max=20),
            reraise=True,
        )
        def _call_ollama(text: str):
            response = requests.post(
                f"{self.ollama_base_url}/api/embeddings",
                json={"model": self.embedding_model, "prompt": text},
                timeout=60,
            )
            response.raise_for_status()
            return response.json()

        for text in texts:
            result = _call_ollama(text)
            embedding = result.get("embedding")
            if embedding is None:
                raise ValueError(f"No embedding found in response: {result}")
            embeddings.append(embedding)

        return embeddings

    def insert_vectors(self, sentences: List[Dict], doc_id: str):
        """
        Insert sentence embeddings into Milvus using MilvusClient.

        Args:
            sentences: List of sentence dictionaries with keys:
                - id: Sentence ID
                - text: Sentence text
                - page_num: Page number
                - paragraph_id: Paragraph ID
            doc_id: Document UUID
        """
        if not sentences:
            return

        # Generate embeddings
        texts = [s["text"] for s in sentences]
        embeddings = self._generate_embeddings(texts)

        # Prepare data as list of dictionaries
        data = []
        for i, sentence in enumerate(sentences):
            data.append(
                {
                    "id": sentence["id"],
                    "embedding": embeddings[i],
                    "doc_id": doc_id,
                    "page_num": sentence["page_num"],
                    "paragraph_id": sentence["paragraph_id"],
                    "sentence_id": sentence["id"],
                }
            )

        # Insert into collection
        self.client.insert(collection_name=self.collection_name, data=data)

        print(f"✓ Inserted {len(sentences)} vectors into Milvus")

    def search(
        self,
        query: str,
        top_k: int = 5,
        filter_doc_id: str = None,
        filter_page_num: int = None,
    ) -> List[Dict]:
        """
        Semantic search using vector similarity with MilvusClient.

        Args:
            query: Search query text
            top_k: Number of results to return
            filter_doc_id: Optional filter by document ID
            filter_page_num: Optional filter by page number

        Returns:
            List of search results with sentence IDs and scores
        """
        # Generate query embedding
        query_embedding = self._generate_embeddings([query])[0]

        # Build filter expression
        filter_expr = None
        if filter_doc_id and filter_page_num:
            filter_expr = (
                f'doc_id == "{filter_doc_id}" && page_num == {filter_page_num}'
            )
        elif filter_doc_id:
            filter_expr = f'doc_id == "{filter_doc_id}"'
        elif filter_page_num:
            filter_expr = f"page_num == {filter_page_num}"

        # Search using MilvusClient
        search_params = {"metric_type": "COSINE", "params": {"nprobe": 10}}

        results = self.client.search(
            collection_name=self.collection_name,
            data=[query_embedding],
            anns_field="embedding",
            search_params=search_params,
            limit=top_k,
            filter=filter_expr,
            output_fields=["doc_id", "page_num", "paragraph_id", "sentence_id"],
        )

        # Format results
        formatted_results = []
        for hit in results[0]:
            formatted_results.append(
                {
                    "sentence_id": hit.get("entity", {}).get("sentence_id")
                    or hit.get("sentence_id"),
                    "score": float(hit.get("distance", 0)),
                    "metadata": {
                        "doc_id": hit.get("entity", {}).get("doc_id")
                        or hit.get("doc_id"),
                        "page_num": hit.get("entity", {}).get("page_num")
                        or hit.get("page_num"),
                        "paragraph_id": hit.get("entity", {}).get("paragraph_id")
                        or hit.get("paragraph_id"),
                    },
                }
            )

        return formatted_results

    def page_exists(self, doc_id: str, page_num: int) -> bool:
        """
        Check if vectors already exist for a specific page.

        Args:
            doc_id: Document UUID
            page_num: Page number

        Returns:
            True if vectors exist for this page, False otherwise
        """
        filter_expr = f'doc_id == "{doc_id}" && page_num == {page_num}'

        try:
            results = self.client.query(
                collection_name=self.collection_name,
                filter=filter_expr,
                limit=1,
                output_fields=["sentence_id"],
            )
            return len(results) > 0
        except Exception as e:
            raise Exception(f"Failed to check page existence: {e}")

    def delete_by_doc_id(self, doc_id: str):
        """
        Delete all vectors for a document using MilvusClient.

        Args:
            doc_id: Document UUID
        """
        self.client.delete(
            collection_name=self.collection_name, filter=f'doc_id == "{doc_id}"'
        )
        print(f"✓ Deleted vectors for doc_id: {doc_id}")

    def close(self):
        """Close Milvus connection."""
        self.client.close()
        print("✓ Milvus connection closed")

File: src/rag/storage/vector-index.ts
import { MilvusClient, DataType } from "@zilliz/milvus2-sdk-node";
import OpenAI from "openai";
import { getMilvusConfig, getEmbeddingConfig } from "../../config/settings.js";
import type { MilvusConfig, EmbeddingConfig } from "../../config/settings.js";

export interface VectorSearchResult {
  readonly sentence_id: string;
  readonly score: number;
  readonly metadata: {
    readonly doc_id: string;
    readonly page_num: number;
    readonly paragraph_id: string;
  };
}

async function withRetry<T>(fn: () => Promise<T>, maxAttempts = 5, baseDelay = 2000): Promise<T> {
  let lastError: unknown;
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (e) {
      lastError = e;
      if (attempt < maxAttempts) {
        const delay = Math.min(baseDelay * Math.pow(2, attempt - 1), 20000);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }
  }
  throw lastError;
}

export class VectorIndex {
  private readonly client: MilvusClient;
  private readonly collectionName: string;
  private readonly dimension: number;
  private readonly embeddingModel: string;
  private readonly provider: string;
  private readonly openaiClient: OpenAI | null;
  private readonly ollamaBaseUrl: string;

  constructor(collectionName = "sentence_embeddings", uri?: string, token?: string) {
    this.collectionName = collectionName;

    const milvusConfig: MilvusConfig = getMilvusConfig();
    const embeddingConfig: EmbeddingConfig = getEmbeddingConfig();

    this.dimension = embeddingConfig.dimension;
    this.embeddingModel = embeddingConfig.model;
    this.provider = embeddingConfig.provider;
    this.ollamaBaseUrl = embeddingConfig.ollamaBaseUrl;

    if (this.provider === "openai") {
      this.openaiClient = new OpenAI({ apiKey: embeddingConfig.openaiApiKey });
    } else {
      this.openaiClient = null;
    }

    const resolvedUri = uri ?? milvusConfig.uri;
    const resolvedToken = token ?? milvusConfig.token;

    this.client = resolvedToken
      ? new MilvusClient({ address: resolvedUri, token: resolvedToken })
      : new MilvusClient({ address: resolvedUri });

    console.log(`[milvus] Connected to ${resolvedUri}`);
  }

  async createCollection(dropExisting = false): Promise<void> {
    if (dropExisting) {
      const exists = await this.client.hasCollection({ collection_name: this.collectionName });
      if (exists.value) {
        await this.client.dropCollection({ collection_name: this.collectionName });
        console.log(`[milvus] Dropped existing collection: ${this.collectionName}`);
      }
    }

    const exists = await this.client.hasCollection({ collection_name: this.collectionName });
    if (exists.value) {
      console.log(`[milvus] Collection already exists: ${this.collectionName}`);
      return;
    }

    await this.client.createCollection({
      collection_name: this.collectionName,
      fields: [
        { name: "id", data_type: DataType.VarChar, max_length: 200, is_primary_key: true },
        { name: "embedding", data_type: DataType.FloatVector, dim: this.dimension },
        { name: "doc_id", data_type: DataType.VarChar, max_length: 100 },
        { name: "page_num", data_type: DataType.Int64 },
        { name: "paragraph_id", data_type: DataType.VarChar, max_length: 100 },
        { name: "sentence_id", data_type: DataType.VarChar, max_length: 100 },
      ],
    });

    await this.client.createIndex({
      collection_name: this.collectionName,
      field_name: "embedding",
      index_type: "IVF_FLAT",
      metric_type: "COSINE",
      params: { nlist: 1024 },
    });

    console.log(`[milvus] Created collection: ${this.collectionName} with COSINE similarity`);
  }

  async generateEmbeddings(texts: readonly string[]): Promise<number[][]> {
    if (this.provider === "ollama") {
      return this.generateOllamaEmbeddings(texts);
    }
    return this.generateOpenAIEmbeddings(texts);
  }

  private async generateOpenAIEmbeddings(texts: readonly string[]): Promise<number[][]> {
    if (!this.openaiClient) throw new Error("OpenAI client not initialized");

    const response = await withRetry(() =>
      this.openaiClient!.embeddings.create({
        model: this.embeddingModel,
        input: texts as string[],
      })
    );

    return response.data.map((item) => item.embedding);
  }

  private async generateOllamaEmbeddings(texts: readonly string[]): Promise<number[][]> {
    return withRetry(async () => {
      const response = await fetch(`${this.ollamaBaseUrl.replace("/v1", "")}/api/embed`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ model: this.embeddingModel, input: texts }),
      });

      if (!response.ok) {
        throw new Error(`Ollama embedding failed: ${response.status} ${response.statusText}`);
      }

      const data = (await response.json()) as { embeddings: number[][] };
      return data.embeddings;
    });
  }

  async insertVectors(sentences: readonly { id: string; text: string; page_num: number; paragraph_id: string }[], docId: string): Promise<void> {
    if (sentences.length === 0) return;

    const texts = sentences.map((s) => s.text);
    const embeddings = await this.generateEmbeddings(texts);

    const data = sentences.map((sentence, i) => ({
      id: sentence.id,
      embedding: embeddings[i],
      doc_id: docId,
      page_num: sentence.page_num,
      paragraph_id: sentence.paragraph_id,
      sentence_id: sentence.id,
    }));

    await this.client.insert({ collection_name: this.collectionName, data });
    console.log(`[milvus] Inserted ${sentences.length} vectors`);
  }

  async search(query: string, topK = 5, filterDocId?: string, filterPageNum?: number): Promise<VectorSearchResult[]> {
    const queryEmbedding = (await this.generateEmbeddings([query]))[0];

    let filterExpr: string | undefined;
    if (filterDocId && filterPageNum !== undefined) {
      filterExpr = `doc_id == "${filterDocId}" && page_num == ${filterPageNum}`;
    } else if (filterDocId) {
      filterExpr = `doc_id == "${filterDocId}"`;
    } else if (filterPageNum !== undefined) {
      filterExpr = `page_num == ${filterPageNum}`;
    }

    const results = await this.client.search({
      collection_name: this.collectionName,
      data: [queryEmbedding],
      anns_field: "embedding",
      params: { nprobe: 10 },
      limit: topK,
      filter: filterExpr,
      output_fields: ["doc_id", "page_num", "paragraph_id", "sentence_id"],
    });

    return (results.results ?? []).map((hit: Record<string, unknown>) => ({
      sentence_id: (hit.sentence_id as string) ?? "",
      score: Number(hit.score ?? hit.distance ?? 0),
      metadata: {
        doc_id: (hit.doc_id as string) ?? "",
        page_num: Number(hit.page_num ?? 0),
        paragraph_id: (hit.paragraph_id as string) ?? "",
      },
    }));
  }

  async pageExists(docId: string, pageNum: number): Promise<boolean> {
    const filterExpr = `doc_id == "${docId}" && page_num == ${pageNum}`;
    const results = await this.client.query({
      collection_name: this.collectionName,
      filter: filterExpr,
      limit: 1,
      output_fields: ["sentence_id"],
    });
    return (results.data?.length ?? 0) > 0;
  }

  async deleteByDocId(docId: string): Promise<void> {
    await this.client.delete({
      collection_name: this.collectionName,
      filter: `doc_id == "${docId}"`,
    });
    console.log(`[milvus] Deleted vectors for doc_id: ${docId}`);
  }

  async close(): Promise<void> {
    await this.client.closeConnection();
    console.log("[milvus] Connection closed");
  }
}
